service/src/main/java/org/cbioportal/service/impl/CacheServiceImpl.java

```diff
@@ -31,4 +44,61 @@ public void clearCaches(boolean clearSpringManagedCache) {
         // version of the r-library and may result in problems in the running instance
         // when flushing the cashes, we do not handle these caches in this service.
     }
+
+    // This evicts keys from the general and static caches when updating/adding/deleting a study.
+    public void clearCachesForStudy(String studyId, List<String> allStudyIds, boolean clearSpringManagedCache) throws CacheOperationException {
+
+        // Flush Spring-managed caches (only when cache strategy has been defined).
+        if (clearSpringManagedCache) {
+            attemptEvictSpringManagedCache(buildRegex(studyId, allStudyIds));
+        }
+        
+        // Flush cache used for user permission evaluation.
+        cacheMapUtil.initializeCacheMemory();
+    }
+    
+    private void attemptEvictSpringManagedCache(String pattern) throws CacheOperationException {
+        try {
+            if (cacheManager != null) {
+                cacheManager.getCacheNames().stream()
+                    .forEach(cacheName -> {
+                        cacheUtils.evictByPattern(cacheName, pattern);
+                    });
+            }
+        } catch (RuntimeException e) {
+            e.printStackTrace();
+            LOG.error("Error while evicting cache." + e.getMessage());
+            throw new CacheOperationException("Error while evicting cache.", e);
+        }
+    }
+    
+    // Regex that selects keys that match id of deleted study
+    // or lacking any study id completely. For example:
+    // ^(?=.*study_id_1).*|^(?!.*study_id_1)(?!.*study_id_2)(?!.*study_id_3).*
+    // https://stackoverflow.com/a/8240998/11651683
+    private String buildRegex(String studyId, List<String> allStudyIds) {
+        
+        // make sure imported studyId is in all studies.
+        List<String> all = new ArrayList<>();
+        all.add(studyId);
```

Comment:
Wait- why? If the regex is "contains `studyId`, or does not contain `any study Id`, haven't you already ensured that the string the regex is matching against does _not_ contain `studyId` by the time the matcher reaches the second part of the regex?

Reply:
No, the rule is delete the key when 1) the key contains `studyId`, or 2) the key contains no study id at all (including not containing `studyId`). The part in the regex before the `|` character relates to 1) and the part after relates to 2).

```
^(?=.*study_id_1).*|^(?!.*study_id_1)(?!.*study_id_2)(?!.*study_id_3).*
```

`^(?=.*study_id_1).*` means: any key that contains _study_id_1_.

`^(?!.*study_id_1)(?!.*study_id_2)(?!.*study_id_3).*` means does not contain any study id.